File: demo/java-demo/src/test/java/com/alibaba/demo/basic/DemoInnerClassTest.java
package com.alibaba.demo.basic;

import com.alibaba.testable.core.annotation.MockMethod;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * 演示对内部类的Mock支持
 * Demonstrate support for mocking invocation inside a inner class
 */
class DemoInnerClassTest {

    public static class Mock {
        @MockMethod(targetClass = DemoInnerClass.class)
        String methodToBeMock() {
            return "MockedCall";
        }
    }

    @Test
    void should_mock_invoke_inside_inner_class() throws Exception {
        DemoInnerClass demo = new DemoInnerClass();
        assertEquals("MockedCall", demo.callInnerDemo());
        assertEquals("MockedCall", demo.callAnonymousInner());
        assertEquals("MockedCall", demo.callLambdaInner());
        assertEquals("MockedCall", new DemoInnerClass.StaticInner().demo());
    }
}

File: demo/kotlin-demo/src/test/kotlin/com/alibaba/demo/basic/DemoInnerClassTest.kt
package com.alibaba.demo.basic

import com.alibaba.testable.core.annotation.MockMethod
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

/**
 * 演示对内部类的Mock支持
 * Demonstrate support for mocking invocation inside a inner class
 */
internal class DemoInnerClassTest {

    class Mock {
        @MockMethod(targetClass = DemoInnerClass::class)
        fun methodToBeMock(): String {
            return "MockedCall"
        }
    }

    @Test
    @Throws(Exception::class)
    fun should_mock_invoke_inside_inner_class() {
        val demo = DemoInnerClass()
        assertEquals("MockedCall", demo.callInnerDemo())
        assertEquals("MockedCall", demo.callAnonymousInner())
        assertEquals("MockedCall", DemoInnerClass.StaticInner().demo())
    }
}
